### ISSUE TITLE
High color images break `img2sixel -I`

### ISSUE THREAD

**Issue by ctrlcctrlv** [libsixel/libsixel#16]
This is a tracking issue against saitoha#61.

This issue has a bit of a storied history already.

* @schauveau opens it in 2017;
* @alxpettit reproduces in 2019;
* @auppal opens #147 in February;
* I merge #1 with the note in saitoha#154 "I found the loss of bgindex in certain functions somewhat troubling, but until there's an API for that I don't think it even matters."
* I revert @auppal (49f43264d1eaf4f6fe267da6a5daa0f2fd1e55cc) because `context.bgimage` is actually useful and can't be removed.

That's where it stands…maybe @auppal will come back and fix it, if not I probably will end up doing so.

**Comment by ctrlcctrlv** [comment]
I have looked more into this issue. The issue is primarily that to implement high color, libsixel writes many palletes to only one image. The decoder doesn't understand this. `SIXEL_COLOR_OPTION_HIGHCOLOR` is only a feature of the encoder. If the decoder comes across another palette in the middle of image data, it considers that we are _rewriting_ 0-255, not *adding new colors*.

This is why @auppal simply said, "remove `bgindex`, make all of this more stateful, where the decoder is writing the color it knows about _at the time_". And indeed, such will work. However, this is not a good solution. For one thing, it allocates 3x more memory for the common case, one palette per Sixel image. For another thing, it destroys entirely the concept of `bgimage` (transparent color key) as we can't choose a transparent color if that color _could_ be used in future by another palette. It also destroys entirely the utility of the palette output from the function.

I see why he did that, but it's an unacceptable solution, as transparency is very important. The memory usage is not so important in comparison.

I will have to think on this problem. Perhaps, a sensible solution is to add a flag to the decoder program `sixel2png`, `-h` for high color, and then, if this is set, write _RGBA_ image (not RGB as @auppal did), following @auppal's method of allocating in this case 4x more memory, but keeping around bgindex and using it _immediately while writing in this mode_ to make a transparent pixel.

I will self-assign, with this understanding that it's the encoder with the issue and not the decoder.

**Comment by dankamongmen** [comment]
when i was testing libsixel, i recall running into issues with the images from AllRGB. i integrated them into the CI for notcurses, and have found them to be useful test cases...hrmm, it seems to be working in alacritty. perhaps the issue was just a small maximum sixel size in xterm?

```
[schwarzgerat](0) $ time img2sixel /media/chungus/images/allrgb/balloon.png 

real    0m12.649s
user    0m12.416s
sys     0m0.156s
[schwarzgerat](0) $ 
```

either way, 12s -- that's a pretty rough time

**Comment by ctrlcctrlv** [comment]
The issue is with the decoder and not the encoder. The decoder doesn't understand that when a Sixel palette entry changes, that doesn't mean that all the times that palette entry was used before change. It only means future uses of it change. It's a pretty big problem and breaking changes to the ABI are needed to fix it, which I'm happy to make. This just cements the idea in my mind that I mentioned in #20, we're going to be maintaining two branches of libsixel for the foreseeable future.

1.9.x will be the legacy release which will receive security updates;
2.0.x will have the improved ABI with fixes to this and #15. Actually, this issue is very intertwined with #15, because the reason transparency doesn't work is similar.

**Comment by ctrlcctrlv** [comment]
Basically I've settled on how I'm going to fix this. The pixels returned are no longer going to be `unsigned char**`. They're going to be `unsigned int**`. `-1` will be transparent/unset. Palette indexes will be available above 255. When a Sixel graphic reassigns one of its color indexes, we'll add a new palette entry like we're supposed to be doing for the final image.

**Comment by ctrlcctrlv** [comment]
Although if we are entertaining ABI changes perhaps dropping entirely this whole silly idea of `sixel_decoder_raw` giving us a palette that we're supposed to apply to images needs to go away, and we just always stick RGBA graphics in `pixels`, keeping `ncolors` around for informational purposes, and dropping `palette`. It'd be much simpler. Thoughts on that @dankamongmen ?

**Comment by ctrlcctrlv** [comment]
Another thing I can do is consider `sixel_decoder_raw` and `sixel_decode_raw` as both being deprecated functions, and add a new function for getting back RGBA data. `sixel_decode_rgba`. That function can support changing palettes, etc.

**Comment by dankamongmen** [comment]
might a `stdint.h` width-specific type be better?

**Comment by dankamongmen** [comment]
> Although if we are entertaining ABI changes perhaps dropping entirely this whole silly idea of `sixel_decoder_raw` giving us a palette that we're supposed to apply to images needs to go away, and we just always stick RGBA graphics in `pixels`, keeping `ncolors` around for informational purposes, and dropping `palette`. It'd be much simpler. Thoughts on that @dankamongmen ?

i would need review the API before i could give a meaningful answer here; i haven't looked at it in detail recently.

**Comment by ctrlcctrlv** [comment]
Certainly, but maybe I should just add a new function rather than expect users to start dealing with palettes above 256 colors. A lot needs to change for us to start throwing palettes multiple MB large at users, I'm afraid.
